File: leetcode/medium/49. Group Anagrams/solution.cc
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> umap;
        vector<vector<string>> res;

        for (string str: strs) {
            string s = str;
            sort(s.begin(), s.end());
            umap[s].push_back(str);
        }

        for (auto it = umap.begin(); it != umap.end(); it++) {
            res.push_back(it->second);
        }

        return res;
    }
};

File: leetcode/medium/49. Group Anagrams/solution.js
/** https://leetcode.com/problems/group-anagrams/
 * @param {string[]} strs
 * @return {string[][]}
 */
function groupAnagrams(strs) {
    const map = new Map();
    const sorted = strs.map((str, i) => [str.split('').sort().join(''), i]);

    sorted.forEach(s => {
        const [val, index] = s;
        map.has(val)
            ? map.set(val, [...map.get(val), strs[index]])
            : map.set(val, [strs[index]]);
    });

    return Array.from(map).map(v => v[1]);
};


groupAnagrams(["", ""])
groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])